src/modules/routing/handlers/BatchHandler.ts

```diff
@@ -0,0 +1,26 @@
+import { EventEmitter } from 'events'
+import { Handler, HandlerInboundMessage } from '../../../agent/Handler'
+
+import { BatchMessage } from '../messages'
+
+export class BatchHandler implements Handler {
+  private eventEmitter: EventEmitter
+  public supportedMessages = [BatchMessage]
+
+  public constructor(eventEmmiter: EventEmitter) {
+    this.eventEmitter = eventEmmiter
+  }
+
+  public async handle(messageContext: HandlerInboundMessage<BatchHandler>) {
+    if (!messageContext.connection) {
+      throw new Error(`Connection for verkey ${messageContext.recipientVerkey} not found!`)
+    }
+
+    const { message } = messageContext
+    const forwardedMessages = message.messages
+
+    forwardedMessages.forEach((message) => {
+      this.eventEmitter.emit('agentMessage', message.message)
+    })
```

Comment:
Do you think it will cause problems that all messages are emitted at the same time? E.g. If we receive 10 messages, they all get processed at the same time, meaning message 10 could be processed before message 5. It could also mean we fetch a record, another received message updated the record in the meantime and then we overwrite the new data.


I'm not sure if we want to take such things into account. A message queue could solve this, but  being able to handle message in parallel would be nice. ACA-Py handles this by fetching a record 'for-update' which means it'll lock other sessions from writing until you are done. Reading is always possible.

Reply:
Yeah. I'm afraid that the problem can occur. My first idea was also a message queue. You're right that it could be convenient to handle messages in parallel in some cases. I'm surprised that ACA-Py decided to go with locking of objects which sounds quite sophisticated to me, but maybe it's the only option 🤷‍♂️